bot/src/modules/level/lib/level.ts

```diff
@@ -0,0 +1,276 @@
+    }
+
+    return array;
+  }
+
+  public static async getLevelData(guildId: string, userId: string) {
+    let levelData = await state.db.level.findUnique({
+      where: {
+        userId_guildId: { guildId, userId },
+      },
+    });
+
+    if (!levelData)
+      levelData = await state.db.level.create({
+        data: {
+          user: connectOrCreate(userId),
+          guild: connectOrCreate(guildId),
+        },
+      });
+
+    return levelData;
+  }
+
+  // Timeout functions
+  public isUserTimedOut(member: GuildMember, timeout = 60 * 1000): boolean {
+    const id = `${member.guild.id}-${member.id}`;
+    const lastTimeout = this.lastXPgrant.get(id);
+    if (lastTimeout && Date.now() - timeout < lastTimeout) return false;
+    this.lastXPgrant.set(id, Date.now());
+    return true;
+  }
+
+  public setUserTimeout(member: GuildMember) {
+    const id = `${member.guild.id}-${member.id}`;
+    this.lastXPgrant.set(id, Date.now());
+  }
+
+  public async clearExpiredTimeouts() {
+    const now = Date.now();
+    const expiryDuration = 10 * 60 * 1000;
+    this.lastXPgrant.forEach((value, key) => {
+      if (now - value > expiryDuration) this.lastXPgrant.delete(key);
+    });
+  }
+
+  // XP functions
+  public async changeUserXP(guildId: string, userId: string, { voiceXP, messageXP }: { voiceXP?: number, messageXP?: number }) {
+    const oldData = await LevelController.getLevelData(guildId, userId);
+    const newData = await state.db.level.update({
+      where: { userId_guildId: { guildId: oldData.guildId, userId: oldData.userId } },
+      data: { messageExperience: { increment: messageXP || 0 }, voiceExperience: { increment: voiceXP || 0 } },
+    });
+
+    const newLevel = this.getLevelFromXP(newData.messageExperience + newData.voiceExperience);
+    const oldLevel = this.getLevelFromXP(oldData.messageExperience + oldData.voiceExperience);
+
+    const levelDelta = newLevel.level - oldLevel.level;
+
+    return { oldData, newData, oldLevel, newLevel, levelDelta };
+  }
+
+  public async vcLoop() {
+    let guilds = state.client.guilds.cache;
+
+    // Filter out guilds that don't have the level system and voice XP gain enabled.
+    guilds = guilds.filter((x) => {
+      const config = state.guilds.get(x.id);
+      return config && config.levelSystemEnabled && config.levelVoiceXPGain;
+    });
+
+    for (const guild of guilds.values()) {
+      const channels = guild.channels.cache.filter((x) => x.type === ChannelType.GuildVoice) as Collection<string, VoiceChannel>;
+
+      for (const channel of channels.values()) {
+        // Filter out bots and deafened users.
+        const members = channel.members
+          .filter(member => !member.user.bot)
+          .filter(member => !member.voice.deaf);
+
+        // filter out users that are alone.
+        if (members.size <= 1) continue;
+
+        for (const member of members.values()) {
+          await this.handleXPEvent(member, null, { xp: 20, timeout: 5 * 60 * 1000 });
```

Comment:
I assume treating voice XP the same as message XP when it comes to timeouts is intentional?

Reply:
yeah I figured this would prevent people from getting double XP and simplifies the state